### PR TITLE
Migrated test_remove_brick_scenarios

### PR DESCRIPTION
**Pull request by aujjwal-redhat**
Migrated the [test
case](https://github.com/gluster/glusto-tests/blob/HEAD/tests/functional/glusterd/test_remove_brick_scenarios.py)

Fixes: #667

Updates: #292

Signed-off-by: Ayush Ujjwal <aujjwal@redhat.com>

### CODE REVIEW

**Comment by aujjwal-redhat** [comment]
```js
✔ Environment setup successful.
✔ Running test case : test_remove_brick_scenarios-dist-rep
✔ test_remove_brick_scenarios-dist-rep Succeeded
Summary:
+---------------+-------+--------------+
|    Category   | Cases | Pass Percent |
+---------------+-------+--------------+
| nonDisruptive |   0   |      0       |
|   Disruptive  |   1   |    100.0     |
|     Total     |   1   |    100.0     |
+---------------+-------+--------------+
Table:
 test_remove_brick_scenarios
 
+-------------+-------------+--------------------+
| Volume Type | Test Result |  Time taken (sec)  |
+-------------+-------------+--------------------+
|   dist-rep  |     PASS    | 2602.0067014694214 |
+-------------+-------------+--------------------+



Framework runtime : 2613.1035239696503

```

File: tests/functional/glusterd/test_remove_brick_scenarios.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+        # Checking status of brick remove operation for other pair of bricks.
+        ret = redant.remove_brick(self.server_list[0], self.vol_name,
+                                  remove_brick_list_other_pair, 'status',
+                                  excep=False)
+        if ret['msg']['opRet'] == 0:
+            raise Exception("Error: Got status on other pair of bricks.")
+
+        # Stopping remove operation for non-existent bricks.
+        remove_brick_list_non_existent = [bricks_list[0] + 'non-existent',
```
Comment (srijan-sivakumar) [inline tests/functional/glusterd/test_remove_brick_scenarios.py:103]: use f strings for string concatenation.

**Review by srijan-sivakumar — CHANGES_REQUESTED** [review]

File: tests/functional/glusterd/test_remove_brick_scenarios.py (hunk 1 above)
Comment (aujjwal-redhat) [reply tests/functional/glusterd/test_remove_brick_scenarios.py:103]: Ack

**Comment by aujjwal-redhat** [comment]
```js
|    Category   | Cases | Pass Percent |
+---------------+-------+--------------+
| nonDisruptive |   0   |      0       |
|   Disruptive  |   1   |    100.0     |
|     Total     |   1   |    100.0     |
+---------------+-------+--------------+
Table:
 test_remove_brick_scenarios
 
+-------------+-------------+--------------------+
| Volume Type | Test Result |  Time taken (sec)  |
+-------------+-------------+--------------------+
|   dist-rep  |     PASS    | 2491.1046924591064 |
+-------------+-------------+--------------------+



Framework runtime : 2501.6858842372894
```

**Comment by srijan-sivakumar** [comment]
2k seconds!!

**Comment by aujjwal-redhat** [comment]
> 2k seconds!!

yeah, the IO operation is quite heavy. It creates 100000 files and I guess gluster takes time in such large number of files.
For less number of files it took less time. Checked with 10, 100, 1000 and 10000 as well.

**Comment by nik-redhat** [comment]
If we use a lesser number of files will that be a problem? If not let's reduce the number to 10000 and see if that works.

**Comment by aujjwal-redhat** [comment]
> If we use a lesser number of files will that be a problem? If not let's reduce the number to 10000 and see if that works.

yeah 10000 works as well and it takes less time. But will that be fine for the test case logic coz 100000 files are also working. Just the time is too much. :-|

**Comment by srijan-sivakumar** [comment]
> > If we use a lesser number of files will that be a problem? If not let's reduce the number to 10000 and see if that works.
> 
> yeah 10000 works as well and it takes less time. But will that be fine for the test case logic coz 100000 files are also working. Just the time is too much. :-|

Technically the idea of creating large number of files is to have file entries under all bricks and not just some bricks. This is because of how gluster works.
1. With the help of hash determined from the file, it is decided which subvolume that file will reside, though in case of the dirs, they are placed on all subvolumes.
2. Coming to layout, which we can simply understand as the range of hashes attached to each brick and files whose hash falls under that range will be stored in those said bricks. Add and remove brick in this way would re-trigger how the layout is assigned to all bricks, hence the need to re-trigger fix-layout which will redo the calculations and assign new layouts, but that'd mean some files will exist in bricks where they're now not supposed to. Now what I'm still unsure of is whether there is a separate migrate phase or would the fix-layout itself take care of the migrating of files. @nik-redhat any idea ?

**Comment by nik-redhat** [comment]
So, we have two ways to trigger rebalance for a volume:
1. `gluster v rebal VOLNAME fix-layout start`: This will only update the layout of the volumes in the process and won't migrate the data.
2. `gluster v rebal VOLNAME start`: This uses the `fix-layout and migration` strategy and does the above task as well as migrates the data based on the storage of the bricks, uniformly.

So, generally what we do is the second option, and the same is being done in our OPs. I think we can go ahead with the TC and run it for 100000 files. What say?

**Comment by srijan-sivakumar** [comment]
> So, we have two ways to trigger rebalance for a volume:
> 
>     1. `gluster v rebal VOLNAME fix-layout start`: This will only update the layout of the volumes in the process and won't migrate the data.
> 
>     2. `gluster v rebal VOLNAME start`: This uses the `fix-layout and migration` strategy and does the above task as well as migrates the data based on the storage of the bricks, uniformly.
> 
> 
> So, generally what we do is the second option, and the same is being done in our OPs. I think we can go ahead with the TC and run it for 100000 files. What say?

As discussed offline ( off-github ), going ahead with this TC.

**Review by srijan-sivakumar — APPROVED** [review]

**Pull request merged**